File: libcpp/src/document.cc
#include <logia/document.hh>

#include <chrono>
#include <cstdlib>
#include <cstring>
#include <errno.h>
#include <iostream>
#include <random>
#include <sstream>
#include <sys/stat.h>
#include <sys/types.h>

#include <logia/doc-config.hh>

#define DATA_DIR (CMAKE_SRC_DIR "/../data")

namespace logia {

namespace {

long get_time_ms() {
  using namespace std::chrono;
  milliseconds ms =
      duration_cast<milliseconds>(system_clock::now().time_since_epoch());
  return ms.count();
}

bool is_legal_id_char(char c) {
  if (c >= 'a' && c <= 'z')
    return true;
  if (c >= 'A' && c <= 'Z')
    return true;
  if (c >= '0' && c <= '9')
    return true;
  return c == '_';
}

std::string gen_uid() {
  constexpr std::size_t UID_SIZE = 24;

  static std::mt19937 gen(std::random_device{}());
  std::uniform_int_distribution<> dist(0, 128);

  std::string res;
  while (res.size() < UID_SIZE) {
    auto c = dist(gen);
    if (is_legal_id_char(c))
      res.push_back(c);
  }

  return res;
}

void create_dir(const std::string &path) {
  if (mkdir(path.c_str(), S_IRWXU) < 0) {
    std::cerr << "MDLogger: Failed to create document directory `" << path
              << "': " << strerror(errno)
              << " (Maybe document ID is already taken ?)\n";
    std::abort();
  }
}

std::string build_data_dir(const std::string &id) {
  auto name = std::string(DATA_DIR) + "/doc_" + id;
  create_dir(name);
  return name;
}

} // namespace

Document::Document(const std::string &name, const std::string &type,
                   bool link_in_root) {
  _conf.id = gen_uid();
  _conf.link_in_root = link_in_root;
  _conf.name = name;
  _conf.time = get_time_ms();
  _conf.type = type;

  _dir = build_data_dir(_conf.id);
  _md_path = _dir + "/main.md";
  _os.open(_md_path);
  _unique = 0;

  auto conf_path = _dir + "/config.txt";
  _conf.write_to_file(conf_path);
}

Document::~Document() {

  if (!_os.good()) {
    std::cerr << "MDLogger: Failed to write markdown file  `" << _md_path << "'"
              << std::endl;
    std::abort();
  }
}

std::string Document::gen_file_name(const std::string &post) {
  return std::to_string(_unique++) + post;
}

std::string Document::gen_file_path(const std::string &post) {
  return _dir + "/" + gen_file_name(post);
}

std::string Document::get_file_path(const std::string &filename) const {
  return "/docs/doc_" + id() + "/" + filename;
}

} // namespace logia

File: libpython/logia/document.py
import os
import random
import time

from .docconfig import DocConfig

DATA_DIR = os.path.join(os.path.dirname(__file__), "../../data")

def get_time_ms():
    return int(time.time() * 1000)

def is_legal_id_char(c):
    if c >= 'a' and c <= 'z':
        return True
    if c >= 'A' and c <= 'Z':
        return True
    if c >= '0' and c <= '9':
        return True
    return c == '_'

def gen_uid():
    UID_SIZE = 24

    res = ""
    while len(res) < UID_SIZE:
        c = chr(random.randint(0, 127))
        if is_legal_id_char(c):
            res += c

    return res
        

def build_data_dir(id):
    name = os.path.join(DATA_DIR, "doc_{}".format(id))
    os.mkdir(name)
    return name

'''
Fromatted text document
Regroups functions common to all documents types
Most of the time this class shouldn't be used directly
Instead a subclass for a specific document type should be used
'''
class Document:

    def __init__(self, name, type, link_in_root):
        self.conf = DocConfig()
        self.conf.id = gen_uid()
        self.conf.link_in_root = link_in_root
        self.conf.name = name
        self.conf.time = get_time_ms()
        self.conf.type = type

        self.dir = build_data_dir(self.conf.id)
        self.md_path = os.path.join(self.dir, "main.md")
        self.os = open(self.md_path, "w")
        self.unique = 0

        self.conf.write2file(os.path.join(self.dir, "config.txt"))

    def __del__(self):
        self.os.close()

    def name(self):
        return self.conf.name

    def id(self):
        return self.conf.id

    def doc_type(self):
        return self.conf.type

    def out_dir(self):
        return self.dir

    def write(self, data):
        if callable(getattr(data, "logia_doc_write", None)):
            data.logia_doc_write(self)
        else:
            self.os.write(data)

    def flush(self):
        self.os.flush()

    # Gen filepath of the form <out_dir>/UID<post>
    def gen_file_name(self, post = ""):
        res = "{}{}".format(self.unique, post)
        self.unique += 1
        return res

    # Gen filepath of the form <out_dir>/UID<post>
    def gen_file_path(self, post = ""):
        return os.path.join(self.dir, self.gen_file_name(post))

    # Get file path of a local file, used on the WebApp to ref this file
    def get_file_path(self, filename):
        return "/docs/doc_{}/{}".format(self.conf.id, filename)
    
